# Исследование выживших на Титанике

## Загрузка библиотек и исследование данных

In [1]:
# импорт библиотек
import pandas as pd

In [2]:
# прочитаем и посмотрим датасет с данными выживших и погибших пасажиров 
gender_submission = pd.read_csv('C:\\titanic\\gender_submission.csv')
# gender_submission.head(20)

### Ислледование датасета train c набором обучающих данных 

In [3]:
# прочитаем и посмотрим датасет c набором обучающих данных 
train = pd.read_csv('C:\\titanic\\train.csv')
train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [4]:
# смотрим сколько всего строк и столбцов в датасете
train.shape

(891, 12)

In [5]:
# смотрим описание средних величин, min/max, среднее, средневкадратичесоке для числительных типов данных
train.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [6]:
# смотрим описание статистику к нечисловым столбцам (категориальному типу)
train.describe(include=['O'])

Name   Sex  Ticket    Cabin Embarked
count                       891   891     891      204      889
unique                      891     2     681      147        3
top     Braund, Mr. Owen Harris  male  347082  B96 B98        S
freq                          1   577       7        4      644

Давайте проанализируем:

Видно, что сущесвуют дубликаты номеров билетов Ticket и переиспользуются каюты Cabins (уникальных записей (unique) меньше, чем общего количества). <br> 
Самый большой порядок дубликата билета - "CA. 2343". Он повторился 7 раз. Аналогично, наибОльшее число людей, занимающих одну и ту же каюту - 4. Они используют каюты "C23 C25 C27".<br>
Также можно заметить, что 644 человека отбыли из порта "S".<br>
Среди 891 записей, 577 были мужчины (Male) и оставшием женщины (Female).<br>

In [7]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


Можно увидеть, что значение Age не задано для большого количества записей.
* Из 891 строк, возраст Age задан лишь для 714 записей.
* Аналогично, Каюты Cabin также пропущены для большого количества записей. Только 204 из 891 записей содержат Cabin значения.

In [8]:
# посмотрим сумму пропущенных значений (NaN)
train.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

* Всего 177 записей с пропущенным возрастом (Age),
* 687 записей с пропущенным значение каюты Cabin
* и для 2 записей не заданы порты отправления Embarked.

### Ислледование датасета test c набором тестовых данных 

In [9]:
# прочитаем и посмотрим датасет с данными для теста модели 
test = pd.read_csv('C:\\titanic\\test.csv')
test.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [10]:
test.shape

(418, 11)

Тестовый датасет содержит 418 строк и 11 столбцов.

Записей в Обучающем наборе = 891

Записей в Тестовом наборе = 418

Всего записей = 891+418 = 1309

Можно увидеть, что около 2/3 от общего количества данных - данные из обучающей выборки (Train)
и около 1/3 всех данных - тестовые (Test).

Survived стобец отсутствует в Тестовом наборе. Мы должны натренировать наш классификатор, используя Тестовый набор и сгенерировать предсказания (Survived) для Тестовых данных.

In [11]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Pclass       418 non-null    int64  
 2   Name         418 non-null    object 
 3   Sex          418 non-null    object 
 4   Age          332 non-null    float64
 5   SibSp        418 non-null    int64  
 6   Parch        418 non-null    int64  
 7   Ticket       418 non-null    object 
 8   Fare         417 non-null    float64
 9   Cabin        91 non-null     object 
 10  Embarked     418 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 36.0+ KB


Есть отсутствующие данные для возраста (Age) и в Тестовом наборе тоже.

Из 418 записей в Тестовом наборе , только в 332 строках присутствует возраст (Age).

Cabin также отсутсвуют в большом колчиестве записей. Только 91 строка из 418 содержать значения для поля Cabin.

In [12]:
# посмотрим и посчитаем в каких столбцах не заполнены данные
test.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

Всего 86 записей с пропущенным Возрастом (Age), <br>
327 записей с пропущенными Каютами (Cabin) <br>
и 1 записьс пропущенной информацией о Цене Fare.<br>

### Заполнение пропусков ###

In [13]:
# находим количество Mr, Mrs, Miss, Master в столбце Name:
total_Mr = 0
total_Mrs = 0
total_Miss = 0
total_Master = 0
for x in train['Name']:
    if "Mr." in x or "Mr " in x:
        total_Mr+=1
    if "Mrs." in x or 'Mrs ' in x:
        total_Mrs +=1
    if "Miss" in x or "Miss." in x:
        total_Miss +=1
    if 'Master' in x or 'Master.' in x:
        total_Master +=1
print(total_Mr + total_Mrs + total_Miss + total_Master)

869


In [14]:
# Находим тех, кто остался в таблице без статуса:
for i in train['Name']:
    if 'Mr.' not in i and "Mrs." not in i and "Miss" not in i and 'Master' not in i and 'Mr ' not in i and 'Mrs ' not in i:
        print(i)

Uruchurtu, Don. Manuel E
Byles, Rev. Thomas Roussel Davids
Bateman, Rev. Robert James
Minahan, Dr. William Edward
Carter, Rev. Ernest Courtenay
Moraweck, Dr. Ernest
Aubart, Mme. Leontine Pauline
Pain, Dr. Alfred
Reynaldo, Ms. Encarnacion
Peuchen, Major. Arthur Godfrey
Butt, Major. Archibald Willingham
Kirkland, Rev. Charles Leonard
Stahelin-Maeglin, Dr. Max
Sagesser, Mlle. Emma
Simonius-Blumer, Col. Oberst Alfons
Frauenthal, Dr. Henry William
Weir, Col. John
Crosby, Capt. Edward Gifford
Rothes, the Countess. of (Lucy Noel Martha Dyer-Edwards)
Brewe, Dr. Arthur Jackson
Leader, Dr. Alice (Farnham)
Reuchlin, Jonkheer. John George
Harper, Rev. John
Montvila, Rev. Juozas


In [15]:
# функция для отделения статуса пассажира
def status_name(Name):
    if "Mr." in Name or 'Mr 'in Name:
        return 'Mr'
    if 'Mrs.' in Name or 'Mrs ' in Name:
        return 'Mrs'
    if 'Miss' in Name or 'Miss ' in Name:
        return 'Miss'
    if 'Master' in Name:
        return 'Master'

In [16]:
# добавляем новый столбец "status_name" статус людей (холост, женат, замужем, не замужем) 
train.insert(loc=4, column = 'status_name', value = train['Name'].apply(status_name))
train.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name status_name     Sex  \
0                            Braund, Mr. Owen Harris          Mr    male   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...         Mrs  female   
2                             Heikkinen, Miss. Laina        Miss  female   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)         Mrs  female   
4                           Allen, Mr. William Henry          Mr    male   
5                                   Moran, Mr. James          Mr    male   
6                            McCarthy, Mr. Timothy J          Mr    male   
7                     Palsson, Master. Gosta Leonard      Master    male   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)         Mrs  female   
9                Nasser, Mrs. Nicholas (Adele Achem)         Mrs  female   

    Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0  22.0      1      0         A/5 21171   7.2500   NaN        S  
1  38.0      1      0          PC 17599  71.2833   C85        C  
2  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
3  35.0      1      0            113803  53.1000  C123        S  
4  35.0      0      0            373450   8.0500   NaN        S  
5   NaN      0      0            330877   8.4583   NaN        Q  
6  54.0      0      0             17463  51.8625   E46        S  
7   2.0      3      1            349909  21.0750   NaN        S  
8  27.0      0      2            347742  11.1333   NaN        S  
9  14.0      1      0            237736  30.0708   NaN        C

In [17]:
#заменим пустые значения стобца "status_name" на unknown
train.loc[train['status_name'].isna(), 'status_name'] = "unknown"

train.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
status_name      0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [18]:
# функция для категоризации пассажиров по возрасту (только для тех, по котрым есть данные)
def age_category(age):
    if age < 10:
        return 'up to 10'
    if age < 20:
        return 'up to 20'
    if age < 30:
        return 'up to 30'
    if age < 40:
        return 'up to 40'
    if age < 50:
        return 'up to 50'
    if age < 60:
        return 'up to 60'
    if age < 70:
        return 'up to 70'
    if age <= 80:
        return 'up to 80'
    return 'up to ???'
    
train.insert(loc=7, column = 'age_category', value = train['Age'].apply(age_category))
train.head(20)

PassengerId  Survived  Pclass  \
0             1         0       3   
1             2         1       1   
2             3         1       3   
3             4         1       1   
4             5         0       3   
5             6         0       3   
6             7         0       1   
7             8         0       3   
8             9         1       3   
9            10         1       2   
10           11         1       3   
11           12         1       1   
12           13         0       3   
13           14         0       3   
14           15         0       3   
15           16         1       2   
16           17         0       3   
17           18         1       2   
18           19         0       3   
19           20         1       3   

                                                 Name status_name     Sex  \
0                             Braund, Mr. Owen Harris          Mr    male   
1   Cumings, Mrs. John Bradley (Florence Briggs Th...         Mrs  female   
2                              Heikkinen, Miss. Laina        Miss  female   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)         Mrs  female   
4                            Allen, Mr. William Henry          Mr    male   
5                                    Moran, Mr. James          Mr    male   
6                             McCarthy, Mr. Timothy J          Mr    male   
7                      Palsson, Master. Gosta Leonard      Master    male   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)         Mrs  female   
9                 Nasser, Mrs. Nicholas (Adele Achem)         Mrs  female   
10                    Sandstrom, Miss. Marguerite Rut        Miss  female   
11                           Bonnell, Miss. Elizabeth        Miss  female   
12                     Saundercock, Mr. William Henry          Mr    male   
13                        Andersson, Mr. Anders Johan          Mr    male   
14               Vestrom, Miss. Hulda Amanda Adolfina        Miss  female   
15                   Hewlett, Mrs. (Mary D Kingcome)          Mrs  female   
16                               Rice, Master. Eugene      Master    male   
17                       Williams, Mr. Charles Eugene          Mr    male   
18  Vander Planke, Mrs. Julius (Emelia Maria Vande...         Mrs  female   
19                            Masselmani, Mrs. Fatima         Mrs  female   

     Age age_category  SibSp  Parch            Ticket     Fare Cabin Embarked  
0   22.0     up to 30      1      0         A/5 21171   7.2500   NaN        S  
1   38.0     up to 40      1      0          PC 17599  71.2833   C85        C  
2   26.0     up to 30      0      0  STON/O2. 3101282   7.9250   NaN        S  
3   35.0     up to 40      1      0            113803  53.1000  C123        S  
4   35.0     up to 40      0      0            373450   8.0500   NaN        S  
5    NaN    up to ???      0      0            330877   8.4583   NaN        Q  
6   54.0     up to 60      0      0             17463  51.8625   E46        S  
7    2.0     up to 10      3      1            349909  21.0750   NaN        S  
8   27.0     up to 30      0      2            347742  11.1333   NaN        S  
9   14.0     up to 20      1      0            237736  30.0708   NaN        C  
10   4.0     up to 10      1      1           PP 9549  16.7000    G6        S  
11  58.0     up to 60      0      0            113783  26.5500  C103        S  
12  20.0     up to 30      0      0         A/5. 2151   8.0500   NaN        S  
13  39.0     up to 40      1      5            347082  31.2750   NaN        S  
14  14.0     up to 20      0      0            350406   7.8542   NaN        S  
15  55.0     up to 60      0      0            248706  16.0000   NaN        S  
16   2.0     up to 10      4      1            382652  29.1250   NaN        Q  
17   NaN    up to ???      0      0            244373  13.0000   NaN        S  
18  31.0     up to 40      1      0            345763  18.0000   NaN        S  
19   NaN

In [19]:
# общее колличество выживших пассажиров по классам
train_survived_1 = train.pivot_table(index=['Pclass'], values=['Survived'], aggfunc=['count', 'sum', 'mean'])
train_survived_1

count      sum      mean
       Survived Survived  Survived
Pclass                            
1           216      136  0.629630
2           184       87  0.472826
3           491      119  0.242363

Всего из 891 пассажира выжило 342 человека, что составляет 38,38% 
Но по пассажирам это значение распределено не равномерно, так больше всего шансов выжить имели пассажиры: 
* 1 класса 63%
* 2 класса 47%
* 3 класса 24%

In [20]:
# общее колличество выживших пассажиров по полу
train_survived_2 = train.pivot_table(index=['Sex','Pclass'], values=['Survived'], aggfunc=['count', 'sum', 'mean'])
# train_survived_2 = train.pivot_table(index=['Sex'], values=['Survived'], aggfunc=['count', 'sum', 'mean'])
train_survived_2

count      sum      mean
              Survived Survived  Survived
Sex    Pclass                            
female 1            94       91  0.968085
       2            76       70  0.921053
       3           144       72  0.500000
male   1           122       45  0.368852
       2           108       17  0.157407
       3           347       47  0.135447

Всего из 891 пассажира самые высокие шансы выжить имели женщины 74% среди мужчин выживших всего 19%
Что невероятно неравномерно, с учтетом того что мужчмн на Титанике было намного больше 577 человек (65% всех пассажиров), 
женщин 314 человек (35% всех пассажиров) Из мужчин выжили 109 человек, среди женщин 233 человека.
По классам тенденция сохраняется, среди пассажиров 1 класса самый большой процент выживших. И самый большой процент выживших среди женщин из любого класса от 1 и по убывающей.

In [22]:
# общее колличество выживших пассажиров по полу с неизвестным возрастом
train_survived_3 = train.pivot_table(index=['Sex','Pclass', 'age_category'], values=['Survived'], aggfunc=['count', 'sum', 'mean'])
train_survived_3

count      sum      mean
                           Survived Survived  Survived
Sex    Pclass age_category                            
female 1      up to 10            1        0  0.000000
              up to 20           13       13  1.000000
              up to 30           16       15  0.937500
              up to 40           27       27  1.000000
              up to 50           13       13  1.000000
              up to 60           12       11  0.916667
              up to 70            3        3  1.000000
              up to ???           9        9  1.000000
       2      up to 10            8        8  1.000000
              up to 20            8        8  1.000000
              up to 30           25       22  0.880000
              up to 40           17       16  0.941176
              up to 50           10        9  0.900000
              up to 60            6        5  0.833333
              up to ???           2        2  1.000000
       3      up to 10           21       11  0.523810
              up to 20           24       13  0.541667
              up to 30           31       15  0.483871
              up to 40           16        7  0.437500
              up to 50            9        0  0.000000
              up to 70            1        1  1.000000
              up to ???          42       25  0.595238
male   1      up to 10            2        2  1.000000
              up to 20            5        2  0.400000
              up to 30           18        9  0.500000
              up to 40           23       12  0.521739
              up to 50           24        9  0.375000
              up to 60           15        4  0.266667
              up to 70           10        1  0.100000
              up to 80            4        1  0.250000
              up to ???          21        5  0.238095
       2      up to 10            9        9  1.000000
              up to 20           10        1  0.100000
              up to 30           28        0  0.000000
              up to 40           31        3  0.096774
              up to 50            8        1  0.125000
              up to 60            9        0  0.000000
              up to 70            3        1  0.333333
              up to 80            1        0  0.000000
              up to ???           9        2  0.222222
       3      up to 10           21        8  0.380952
              up to 20           42        4  0.095238
              up to 30          102       16  0.156863
              up to 40           53        8  0.150943
              up to 50           25        2  0.080000
              up to 60            6        0  0.000000
              up to 70            2        0  0.000000
              up to 80            2        0  0.000000
              up to ???          94        9  0.095745

In [ ]:
# сводная таблица выживших пассажиров по категориям по возрасту
train_pivot1 = train.pivot_table(index=['Sex', 'Pclass', 'age_category'], values=['Survived'], aggfunc=['count', 'sum', 'mean'])
train_pivot1

In [ ]:
# сводная таблица выживших пассажиров по категориям по возрасту
train_pivot2 = train.pivot_table(index=['Sex', 'status_name', 'age_category'], values=['Survived'], aggfunc=['count', 'sum', 'mean'])
train_pivot2

In [ ]:
# проверяем количество пассажиров после категоризации по возрастам
train_pivot2_count = train_pivot2['count','Survived'].sum()
F'проверяем количество пассажиров после категоризации по возрастам: {train_pivot2_count} из 891'

In [ ]:
# проверяем количество выживших пассажиров после категоризации по возрастам
train_pivot2_sum = train_pivot2['sum','Survived'].sum()
F'проверяем количество выживших пассажиров после категоризации по возрастам: {train_pivot2_sum} из 342'

In [ ]:
train['Age'].max()

In [ ]:
train['Embarked'].unique()

In [ ]:
# составим список пассажиров-детей
